### ISSUE TITLE
PRs are not always correctly closed as merged

### ISSUE THREAD

**Issue by fatho** [channable/hoff#196]
Occasionally, it happens that a PR merged by hoff ends up showing as closed rather than merged, i.e.
![image](https://user-images.githubusercontent.com/3857449/214841497-a186ac97-0e21-4a29-a65c-61e93b905dde.png)
rather than
![image](https://user-images.githubusercontent.com/3857449/214841587-38164d82-30d1-4e32-ba71-8080cbd332cd.png)

And sometimes, github just doesn't close the PR at all.

In hoff, the merging procedure is defined here:

https://github.com/channable/hoff/blob/46805f05a0c23ee1bac5be786da1acd23d076ce3/src/Logic.hs#L203-L210

We first force-push the rebased version of the original branch, and then push the merge to master.

That way, GitHub *should* see that the commits in the PR branch are exactly those that are on master, and correctly close the PR as merged. This is just like a human would do a rebase & merge via the `git` command line (i.e. not using GitHub's interface for that), albeit a lot faster.

The latter point might actually be the issue here. My hypothesis is that if the force push of the PR branch and the push to master are sent in rapid succession, GitHub might become confused about the state of the branches, and incorrectly observe a state where the commits that were merged are not yet on the PR branch.

A hacky band-aid could be to just sleep for a few seconds between force-pushing the PR branch and pushing master.

**Comment by HugoPeters1024** [comment]
Fixed by #209, which ironically was itself affected by the bug

**Comment by fatho** [comment]
I was thinking, we already get webhook events from GitHub when the commit corresponding to a PR changes.
In most cases, that causes hoff to abandon the PR, but we already specifically check whether the updated commit SHA corresponds to the commit we pushed ourselves for integrating the PR.

So maybe we could make hoff wait for that webhook after force-pushing the branch, and only upon receiving it continue with the push of the merge commit to master.

Perhaps we also need some sort of timeout there, too, after which we just proceed to push master anyway. Otherwise we have yet another webhook on the critical path and thus increase our chance of stalls when GitHub has webhook issues.
